src/os/bluestore/BlueStore.cc

```diff
@@ -8827,6 +8827,131 @@ bool BlueStore::has_null_manager() const
   return (fm && fm->is_null_manager());
 }
 
+int BlueStore::mount_readonly()
+{
+  int r = _mount_readonly();
+  if (r < 0) {
+    return r;
+  }
+  r = _open_collections();
+  if (r < 0) {
+    return r;
+  }
+  auto shutdown_cache = make_scope_guard([&] {
+    if (!mounted) {
+      _shutdown_cache();
+    }
+  });
+
+  _kv_start();
+  auto stop_kv = make_scope_guard([&] {
+    if (!mounted) {
+      _kv_stop();
+    }
+  });
+
+  r = _deferred_replay();
```

Comment:
I have a serious doubt if `_deferred_replay()` is compatible with read-only.
In kv-sync-thread we have:
```
for (auto b : deferred_stable) {
....
	  synct->rm_single_key(PREFIX_DEFERRED, key);
....	  
      }
```
My claim is that if there were any deferreds, we will fail.
So maybe it is better not to do it at all?

Reply:
submit_transaction_sync in kv_sync_thread is protected with db_was_opened_read_only flag now: 

int r = db_was_opened_read_only || cct->_conf->bluestore_debug_omit_kv_commit ?
	0 : db->submit_transaction_sync(synct);